pypeit/spectrographs/spectrograph.py

```diff
@@ -790,26 +793,36 @@ def get_meta_value(self, inp, meta_key, required=False, ignore_bad_header=False,
                 msgs.warn("Requested meta data for meta_key={} does not exist...".format(meta_key))
                 return None
 
+        # Is this meta required for this frame type (Spectrograph specific)
+        if ('required_ftypes' in self.meta[meta_key]) and (usr_row is not None):
+            required = False
+            for ftype in self.meta[meta_key]['required_ftypes']:
+                if ftype in usr_row['frametype']:
+                    required = True
+
         # Check if this meta key is required
         if 'required' in self.meta[meta_key].keys():
             required = self.meta[meta_key]['required']
 
         # Is this not derivable?  If so, use the default
         #   or search for it as a compound method
         value = None
-        if self.meta[meta_key]['card'] is None:
-            if 'default' in self.meta[meta_key].keys():
-                value = self.meta[meta_key]['default']
-            elif 'compound' in self.meta[meta_key].keys():
-                value = self.compound_meta(headarr, meta_key)
+        try:
+            if self.meta[meta_key]['card'] is None:
+                if 'default' in self.meta[meta_key].keys():
+                    value = self.meta[meta_key]['default']
+                elif 'compound' in self.meta[meta_key].keys():
+                    value = self.compound_meta(headarr, meta_key)
+                else:
+                    msgs.error("Failed to load spectrograph value for meta: {}".format(meta_key))
             else:
-                msgs.error("Failed to load spectrograph value for meta: {}".format(meta_key))
-        else:
-            # Grab from the header, if we can
-            try:
+                # Grab from the header, if we can
                 value = headarr[self.meta[meta_key]['ext']][self.meta[meta_key]['card']]
-            except (KeyError, TypeError):
-                value = None
+        except (KeyError, TypeError) as e:
+            if ignore_bad_header or (not required):
+                msgs.warn("Bad Header, but we'll try to continue on..") 
+            else:
+                raise e
```

Comment:
Please consider whether we can remove the try: except blocks below.

Reply:
Which lines of code in the block above can fail due to KeyErrror or TypeError? Looks to me like it is only line 820 and possibly also line 815. If compound_meta is parsing headers you should just be passing ignore_bad_header to that routine as well, and then there can be a try:except at exactly the line that can fail. That makes the code easier to read and also easier to debug.